lib/que/adapters/active_record_with_lock.rb

```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+# https://github.com/que-rb/que/blob/80d6067861a41766c3adb7e29b230ce93d94c8a4/lib/que/active_job/extensions.rb
+module Que
+    module Adapters
+      class ActiveRecordWithLock < Que::Adapters::ActiveRecord
+        attr_accessor :job_connection_pool, :lock_connection
+        def initialize(job_connection_pool:, lock_connection:)
+            @job_connection_pool = job_connection_pool
+            @lock_connection = lock_connection
+            super
+        end
+
+        def checkout_activerecord_adapter(&block)
+            @job_connection_pool.with_connection(&block)
+        end
+
+        def lock_database_connection
+            Thread.current[:db_connection] ||= @lock_connection.connection
```

Comment:
My understanding is that the reason we are doing this is so that this works for Que workers that are using multiple threads. Each worker will get it's _own_ connection, meaning that it is never blocked on another thread's work. Is that right?

Alternatively, how does this work with the connection enters a bad state? Does any cleanup need to happen when unlocking a job so that the next lock operation can use a fresh connection?

Reply:
This is mostly for the advisory locks when multiple threads are running. We need the same connection to release the advisory lock; otherwise, it will throw an exception saying the connection does not have permission to release.  Storing this in a thread variable will make sure we are using the same connection to release.

Although now the specs are working fine without storing these as well. 
This might be because previously I was using connection_pool and if any other connection from the connection_pool was trying to release the locks it error. 

But I'm wondering for thread safety If I should use connection_pool instead of .connection 🤔 

>Alternatively, how does this work with the connection enters a bad state?

These edge cases are not yet handled. It will be done in the follow up PR's